src/SiPixelLorentzAngle.cc

```diff
@@ -170,6 +153,57 @@ void SiPixelLorentzAngle::beginJob()
 //   edm::ESHandle<TrackerGeometry> estracker; //this block should not be in beginJob()
 //   c.get<TrackerDigiGeometryRecord>().get(estracker);
 //   tracker=&(* estracker);
+	
+  nlay = 0;
+  for (int i = 0; i < 4; i++){
+     nModules_[i] = 0;
+  }
+}
+
+void SiPixelLorentzAngle::beginRun(const edm::Run & iRun, const edm::EventSetup & es){
+	
+    edm::ESHandle<TrackerTopology> tTopoHandle;
+    es.get<TrackerTopologyRcd>().get(tTopoHandle);
+    const TrackerTopology* const tTopo = tTopoHandle.product();
+    
+    edm::ESHandle<TrackerGeometry> estracker;
+    es.get<TrackerDigiGeometryRecord>().get(estracker);
+    const TrackerGeometry * const geom = &(* estracker);
+    
+    PixelTopologyMap map = PixelTopologyMap(geom,tTopo);
+    
+    if ((geom->isThere(GeomDetEnumerators::P2PXB)) || (geom->isThere(GeomDetEnumerators::P2PXEC))) {
+        edm::LogPrint("PixelTopologyMapTest") << "===== Testing Phase-2 Pixel Tracker geometry =====" << std::endl;
+        nlay = 4;
+    } else if ((geom->isThere(GeomDetEnumerators::P1PXB)) || (geom->isThere(GeomDetEnumerators::P1PXEC))) {
+        edm::LogPrint("PixelTopologyMapTest") << "===== Testing Phase-1 Pixel Tracker geometry =====" << std::endl;
+        nlay = 4;
+    } else {
+        edm::LogPrint("PixelTopologyMapTest") << "===== Testing Phase-0 Pixel Tracker geometry =====" << std::endl;
+        nlay = 3;
+    }
+    
+    for (int i = 0; i < nlay; i++){
+       nModules_[i] = map.getPXBModules(i + 1);
+    }
+	
+	
+  //book histograms
+  char name[128];
+   for(int i_layer = 1; i_layer<=nlay; i_layer++){
+      for(int i_module = 1; i_module<=nModules_[i_layer - 1]; i_module++){
+          sprintf(name, "h_drift_depth_adc_layer%i_module%i", i_layer, i_module);
+          _h_drift_depth_adc_[i_module + (i_layer -1) * 8] = new TH2F(name,name,hist_drift_ , min_drift_, max_drift_, hist_depth_, min_depth_, max_depth_);
+          sprintf(name, "h_drift_depth_adc2_layer%i_module%i", i_layer, i_module);
+          _h_drift_depth_adc2_[i_module + (i_layer -1) * 8] = new TH2F(name,name,hist_drift_ , min_drift_, max_drift_, hist_depth_, min_depth_, max_depth_);
+          sprintf(name, "h_drift_depth_noadc_layer%i_module%i", i_layer, i_module);
+          _h_drift_depth_noadc_[i_module + (i_layer -1) * 8] = new TH2F(name,name,hist_drift_ , min_drift_, max_drift_, hist_depth_, min_depth_, max_depth_);
+          sprintf(name, "h_drift_depth_layer%i_module%i", i_layer, i_module);
+          _h_drift_depth_[i_module + (i_layer -1) * 8] = new TH2F(name,name,hist_drift_ , min_drift_, max_drift_, hist_depth_, min_depth_, max_depth_);
+          sprintf(name, "h_mean_layer%i_module%i", i_layer, i_module);
+          _h_mean_[i_module + (i_layer -1) * 8] = new TH1F(name,name,hist_depth_, min_depth_, max_depth_);
```

Comment:
@wweiphy the multiplication with 8 should be changed, that 8 comes from the hardcoded module numbers, so that should be the `nModules_[i_layer - 1]` I think